crates/torin/src/values/size.rs

```diff
@@ -193,50 +196,79 @@ impl std::fmt::Display for DynamicCalculation {
     }
 }
 
-/// Calculate some chained operations with a given value.
-/// This value could be for example the width of a node's parent area.
-pub fn run_calculations(calcs: &[DynamicCalculation], value: f32) -> f32 {
-    let mut prev_number: Option<f32> = None;
-    let mut prev_op: Option<DynamicCalculation> = None;
+struct DynamicCalculationEvaluator<'a> {
+    calcs: Iter<'a, DynamicCalculation>,
+    parent_value: f32,
+    current: Option<&'a DynamicCalculation>,
+}
+
+impl<'a> DynamicCalculationEvaluator<'a> {
+    pub fn new(calcs: Iter<'a, DynamicCalculation>, parent_value: f32) -> Self {
+        Self {
+            calcs,
+            parent_value,
+            current: None,
+        }
+    }
+
+    pub fn evaluate(&mut self) -> Option<f32> {
+        self.current = self.calcs.next();
+        self.expr(0)
+    }
+
+    pub fn expr(&mut self, precedence: usize) -> Option<f32> {
+        let mut lhs = self.term()?;
+
+        while let Some(p) = self.operator_precedence() {
+            if p < precedence {
+                return Some(lhs);
+            }
+
+            let op = self.current?;
+            self.current = self.calcs.next();
+            let rhs = self.expr(p + 1)?;
 
-    let mut calc_with_op = |val: f32, prev_op: Option<DynamicCalculation>| {
-        if let Some(op) = prev_op {
             match op {
-                DynamicCalculation::Sub => {
-                    prev_number = Some(prev_number.unwrap() - val);
-                }
-                DynamicCalculation::Add => {
-                    prev_number = Some(prev_number.unwrap() + val);
-                }
-                DynamicCalculation::Mul => {
-                    prev_number = Some(prev_number.unwrap() * val);
-                }
-                DynamicCalculation::Div => {
-                    prev_number = Some(prev_number.unwrap() / val);
-                }
-                _ => {}
+                DynamicCalculation::Add => lhs += rhs,
+                DynamicCalculation::Sub => lhs -= rhs,
+                DynamicCalculation::Mul => lhs *= rhs,
+                DynamicCalculation::Div => lhs /= rhs,
+                _ => return None,
             }
-        } else {
-            prev_number = Some(val);
         }
-    };
 
-    for calc in calcs {
-        match calc {
-            DynamicCalculation::Percentage(per) => {
-                let val = (value / 100.0 * per).round();
+        if self.current.is_some() {
+            return None;
+        }
 
-                calc_with_op(val, prev_op);
+        Some(lhs)
+    }
 
-                prev_op = None;
+    pub fn term(&mut self) -> Option<f32> {
```

Comment:
Same for this, "term" doesn't really explain much

Reply:
Would calling it `value` be better? Since currently it's either percentage or pixels. I have seen it being called atom in a lot of grammars and parsers, but I think value would be more intuitive? And would renaming function names to `parse_value` and `parse_expression` be fine?